### ISSUE TITLE
    panic: fatal: A failure has occurred: Unrecognized structpb value kind in RPC[tf.Provider[aws].Invoke(aws:iam/getPolicyDocument:getPolicyDocument).args]: <nil>

### ISSUE THREAD

**Issue by maratsh** [pulumi/pulumi#5236]
```
const policyDoc = pulumi.output(
      aws.iam.getPolicyDocument(
        {
          statements: [
            {
              actions: [
                'sqs:ChangeMessageVisibility',
                'sqs:ChangeMessageVisibilityBatch',
                'sqs:DeleteMessage',
                'sqs:DeleteMessageBatch',
                'sqs:GetQueueAttributes',
                'sqs:ReceiveMessage',
              ],
              resources: [ args.sqsArn ],
            },
          ],
        }
      )
    );

    const policy = new aws.iam.Policy(name, {
      policy: policyDoc.json
    });
```

is throwing on pulumi up

Diagnostics:
  aws:sns:TopicSubscription (orders):
    error: transport is closing
 
  pulumi:pulumi:Stack (stack):
    panic: fatal: A failure has occurred: Unrecognized structpb value kind in RPC[tf.Provider[aws].Invoke(aws:iam/getPolicyDocument:getPolicyDocument).args]: <nil>
    goroutine 97 [running]:
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.failfast(...)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.Failf(0x5fcabac, 0x2f, 0xc001b5cec8, 0x2, 0x2)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/util/contract/fail.go:30 +0x181
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc000c4bd00, 0xc00114c050, 0x49, 0x101, 0xc000766090, 0x0, 0x0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:353 +0x3bd
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc000c4bc80, 0xc00114c050, 0x49, 0x101, 0xc0010228b0, 0x0, 0x0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:271 +0x1091
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalProperties(0xc000fd5e30, 0xc00114c050, 0x49, 0x101, 0x5498c00, 0xc000ffff20, 0x73)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:221 +0x263
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc000c4ba00, 0xc00114c050, 0x49, 0x101, 0x203000, 0x203000, 0x203000)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:286 +0x446
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc000c4b980, 0xc00114c050, 0x49, 0x101, 0xc001022978, 0xc000ffff50, 0x0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:271 +0x1091
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalProperties(0xc000fd5dd0, 0xc00114c050, 0x49, 0x101, 0x1, 0xc00114c050, 0x49)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:221 +0x263
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfbridge.(*Provider).Invoke(0xc000e508c0, 0x6a97a00, 0xc000fd5e90, 0xc00103a7e0, 0xc000e508c0, 0x5295201, 0xc000c4bd40)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.7.3/pkg/tfbridge/provider.go:1064 +0x3b5
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Invoke_Handler.func1(0x6a97a00, 0xc000fd5e90, 0x5cc5640, 0xc00103a7e0, 0x5d19580, 0xaa6b360, 0x6a97a00, 0xc000fd5e90)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/proto/go/provider.pb.go:1809 +0x86
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x6a97a00, 0xc000fd5b90, 0x5cc5640, 0xc00103a7e0, 0xc000c338a0, 0xc000c338c0, 0x0, 0x0, 0x69d9820, 0xc000364700)
    	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2eb
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Invoke_Handler(0x5ddd380, 0xc000e508c0, 0x6a97a00, 0xc000fd5b90, 0xc00103a780, 0xc000e21720, 0x6a97a00, 0xc000fd5b90, 0xc000460480, 0x112)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/proto/go/provider.pb.go:1811 +0x14b
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc000e50700, 0x6abc2e0, 0xc000f38180, 0xc0005b8600, 0xc000ef3530, 0xaa2d4a0, 0x0, 0x0, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171 +0x50a
    google.golang.org/grpc.(*Server).handleStream(0xc000e50700, 0x6abc2e0, 0xc000f38180, 0xc0005b8600, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494 +0xccd
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000772090, 0xc000e50700, 0x6abc2e0, 0xc000f38180, 0xc0005b8600)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834 +0xa1
    created by google.golang.org/grpc.(*Server).serveStreams.func1
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:832 +0x204

## Expected Behavior

getPolicyDocument should not throw errors
## Current Behavior

throwing 
    panic: fatal: A failure has occurred: Unrecognized structpb value kind in RPC[tf.Provider[aws].Invoke(aws:iam/getPolicyDocument:getPolicyDocument).args]: <nil>
    goroutine 97 [running]:

pulumi version
v2.9.0

cat package.json 
{
  "name": "logistic-app",
  "devDependencies": {
    "@types/node": "^10.17.28"
  },
  "scripts": {
    "lint": "eslint **/*.ts --fix"
  },
  "dependencies": {
    "@pulumi/aws": "^3.0.0",
    "@pulumi/pulumi": "^2.9.0",
    "@typescript-eslint/eslint-plugin": "^2.33.0",
    "@typescript-eslint/parser": "^2.33.0",
    "eslint": "^7.7.0",
    "eslint-config-airbnb-base": "^14.1.0",
    "eslint-config-prettier": "^6.11.0",
    "eslint-plugin-import": "^2.22.0",
    "eslint-plugin-prettier": "^3.1.3",
    "prettier": "^2.1.1"
  }
}

## Steps to Reproduce

1.
2.
3.
4.

## Context (Environment)

I'm trying to modify stack, This piece of code was not touched.

**Comment by haidaraM** [comment]
I have the same issue in Python 3.8 with Pulumi v2.10.0. Just run `pulumi up` with following snippet:
```python
from pulumi_aws import iam, s3

jenkins_conf_s3_bucket = s3.Bucket(
    "jenkins-jcasc",
    force_destroy=True,
    acl="private",
    versioning=s3.BucketVersioningArgs(enabled=True),
)

master_ecs_task_policy_document = iam.get_policy_document(
    statements=[
        iam.GetPolicyDocumentStatementArgs(
            sid="S3AccessForJCasC",
            effect="Allow",
            actions=["s3:GetObject"],
            resources=[jenkins_conf_s3_bucket.arn.apply(lambda arn: f"{arn}/*")],
        ),
    ]
)
```

<details><summary>Stack trace: </summary>
<p>

```
pulumi:pulumi:Stack (my-stack-dev):
    panic: fatal: A failure has occurred: Unrecognized structpb value kind in RPC[tf.Provider[aws].Invoke(aws:iam/getPolicyDocument:getPolicyDocument).args]: <nil>
    goroutine 114 [running]:
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.failfast(...)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.Failf(0x6bc984f, 0x2f, 0xc0010daec8, 0x2, 0x2)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/util/contract/fail.go:30 +0x181
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc001374240, 0xc0006c09b0, 0x49, 0x101, 0x101, 0x0, 0x0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:353 +0x3bd
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc0013741c0, 0xc0006c09b0, 0x49, 0x101, 0xc000eec238, 0x0, 0x0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:271 +0x1091
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalProperties(0xc001360bd0, 0xc0006c09b0, 0x49, 0x101, 0x60980c0, 0xc000a99360, 0x73)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:221 +0x263
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc001374080, 0xc0006c09b0, 0x49, 0x101, 0x203000, 0x203000, 0x203000)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:286 +0x446
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalPropertyValue(0xc000031fc0, 0xc0006c09b0, 0x49, 0x101, 0xc000f06088, 0xc000a993d0, 0x0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:271 +0x1091
    github.com/pulumi/pulumi/sdk/v2/go/common/resource/plugin.UnmarshalProperties(0xc0013609f0, 0xc0006c09b0, 0x49, 0x101, 0x1, 0xc0006c09b0, 0x49)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/go/common/resource/plugin/rpc.go:221 +0x263
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfbridge.(*Provider).Invoke(0xc0008a2c40, 0x7697580, 0xc001360c30, 0xc000c2baa0, 0xc0008a2c40, 0x5e94201, 0xc0013742c0)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.7.3/pkg/tfbridge/provider.go:1064 +0x3b5
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Invoke_Handler.func1(0x7697580, 0xc001360c30, 0x68c46c0, 0xc000c2baa0, 0x6918840, 0xb66e018, 0x7697580, 0xc001360c30)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/proto/go/provider.pb.go:1809 +0x86
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x7697580, 0xc0013607b0, 0x68c46c0, 0xc000c2baa0, 0xc000bfd4e0, 0xc000bfd500, 0x0, 0x0, 0x75d9300, 0xc000285d90)
    	/home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2eb
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Invoke_Handler(0x69dca20, 0xc0008a2c40, 0x7697580, 0xc0013607b0, 0xc000c2b9e0, 0xc000fdea40, 0x7697580, 0xc0013607b0, 0xc001340580, 0xa5)
    	/home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.9.1-0.20200825190708-910aa96016cd/proto/go/provider.pb.go:1811 +0x14b
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc0008a2a80, 0x76bc0c0, 0xc000f93200, 0xc00064c500, 0xc00106aae0, 0xb630600, 0x0, 0x0, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171 +0x50a
    google.golang.org/grpc.(*Server).handleStream(0xc0008a2a80, 0x76bc0c0, 0xc000f93200, 0xc00064c500, 0x0)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494 +0xccd
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0010421a0, 0xc0008a2a80, 0x76bc0c0, 0xc000f93200, 0xc00064c500)
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834 +0xa1
    created by google.golang.org/grpc.(*Server).serveStreams.func1
    	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:832 +0x204

    error: Program failed with an unhandled exception:
    error: Traceback (most recent call last):
      File "/usr/local/bin/pulumi-language-python-exec", line 85, in <module>
        loop.run_until_complete(coro)
      File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
        return future.result()
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 83, in run_in_stack
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 51, in run_pulumi_func
        await RPC_MANAGER.rpcs.pop()
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 35, in run_pulumi_func
        func()
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 83, in <lambda>
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 106, in __init__
        func()
      File "/usr/local/bin/pulumi-language-python-exec", line 84, in <lambda>
        coro = pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__'))
      File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 282, in run_path
        return _run_code(code, mod_globals, init_globals,
      File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/runpy.py", line 87, in _run_code
        exec(code, run_globals)
      File "./__main__.py", line 10, in <module>
        master_ecs_task_policy_document = iam.get_policy_document(
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi_aws/iam/get_policy_document.py", line 324, in get_policy_document
        __ret__ = pulumi.runtime.invoke('aws:iam/getPolicyDocument:getPolicyDocument', __args__, opts=opts, typ=GetPolicyDocumentResult).value
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/invoke.py", line 133, in invoke
        return InvokeResult(_sync_await(asyncio.ensure_future(do_rpc())))
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/sync_await.py", line 95, in _sync_await
        return fut.result()
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/invoke.py", line 130, in do_rpc
        raise exn
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/rpc_manager.py", line 67, in rpc_wrapper
        result = await rpc
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/invoke.py", line 112, in do_invoke
        resp = await asyncio.get_event_loop().run_in_executor(None, do_invoke)
      File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/concurrent/futures/thread.py", line 57, in run
        result = self.fn(*self.args, **self.kwargs)
      File "/Users/mhaidara/dev/perso/my-stack/venv/lib/python3.8/site-packages/pulumi/runtime/invoke.py", line 110, in do_invoke
        raise Exception(details)
    Exception: invocation of aws:iam/getPolicyDocument:getPolicyDocument returned an error: transport is closing
    error: an unhandled error occurred: Program exited with non-zero exit code: 1
```

</p>
</details>


<details><summary>Pip freeze: </summary>
<p>

```
Arpeggio==1.9.2
attrs==19.3.0
dill==0.3.2
grpcio==1.31.0
packaging==20.4
parver==0.3.0
protobuf==3.12.4
pulumi==2.10.0
pulumi-aws==3.2.1
pyparsing==2.4.7
semver==2.10.2
six==1.15.0
```
</p>
</details>

**Comment by Sodki** [comment]
I don't know why, but it started happening to me today.

Pulumi v2.13.2
pulumi-aws v3.12.1

It seems to be failing only when Outputs are involved and only when dealing with new resources. It doesn't complain if the resources already exist. To clarify, I'm using `pulumi.aws.get_policy_document()` in a lot of places, but I only complains about new code. Current existing code is not cause for error.

This is good news because I can then workaround those, and only those, with: `pulumi.Output.all(...).apply(lambda vars: json.dumps(...)`. Still, very annoying.

**Comment by stack72** [comment]
So after speaking with @Sodki I am able to see that the following is an issue:


```py
policy_document = aws.iam.get_policy_document(statements=[
    {
        "actions": ["s3:ListBucket"],
        "resources": [bucket.arn]
	},
    {
        "actions": ["s3:PutObject"],
        "resources": [pulumi.Output.concat(bucket.arn, "/*")]
	}])
```

But this works as expected:

```py
policy_document = pulumi.Output.all(bucket.arn).apply(
          lambda vars: aws.iam.get_policy_document(statements=[
            {
                "actions": ["s3:ListBucket"],
                "resources": [vars[0]]
            },
            {
                "actions": ["s3:PutObject"],
                "resources": [f"{vars[0]}/*"]
            }]))
```

the crux of this issue is the resolution of the outputs as part of the datasource - when it's wrapped with output.all, all works as expected

**Comment by dokuboyejo** [comment]
Any update on this...getting the same error

```
policy_document = pulumi.Output.all(bucket_arn).apply(
    lambda args: aws.iam.get_policy_document(
        version='2012-10-17',
        # version='2008-10-17',
        policy_id='PolicyForCloudFrontPrivateContent',
        statements=[
            aws.iam.GetPolicyDocumentStatementArgs(
                sid=1,
                effect='Allow',
                actions=["s3:GetObject"],
                resources=[f'{args[0]}/*'],
                principals=[
                    aws.iam.GetPolicyDocumentStatementPrincipalArgs(
                        type='AWS',
                        identifiers=[origin_access_identity.iam_arn],
                    )
                ],
            )
        ],
        opts=ResourceOptions(depends_on=[cf])
    )
)
```


```
 panic: fatal: A failure has occurred: Unrecognized structpb value kind in RPC[tf.Provider[aws].Invoke(aws:iam/getPolicyDocument:getPolicyDocument).args]: <nil>
    goroutine 79 [running]:
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/util/contract/failfast.go:23
    github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Failf(0x71c5a87, 0x2f, 0xc001fcc308, 0x2, 0x2)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/util/contract/fail.go:30 +0x18d
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalPropertyValue(0xc0005529c0, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0x49ec00, 0x7b987a0, 0xc001fcc470)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:427 +0x3b9
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalPropertyValue(0xc000552940, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0xc0024895e0, 0x1, 0x2)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:292 +0x19c9
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalProperties(0xc002877f20, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:243 +0x267
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalPropertyValue(0xc000552900, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:307 +0x43a
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalPropertyValue(0xc000552880, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0xc00284fd30, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:292 +0x19c9
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalProperties(0xc002877ec0, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0x1135, 0xc001fccfd0, 0x4189f3)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:243 +0x267
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalPropertyValue(0xc000552640, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0xc001fcd230, 0xc001fcd450, 0x4f402c)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:307 +0x43a
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalPropertyValue(0xc0005525c0, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0xc00284fc80, 0x0, 0x1)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:292 +0x19c9
    github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin.UnmarshalProperties(0xc002877e60, 0xc001f9c0a0, 0x49, 0x101, 0x0, 0xc001f9c0a0, 0x49, 0x724ca28)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/go/common/resource/plugin/rpc.go:243 +0x267
    github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge.(*Provider).Invoke(0xc00000ab40, 0x7c59ac8, 0xc002877f80, 0xc0024e1200, 0xc00000ab40, 0x626b601, 0xc000552a40)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v3@v3.0.0/pkg/tfbridge/provider.go:1128 +0x3e5
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Invoke_Handler.func1(0x7c59ac8, 0xc002877f80, 0x6e54980, 0xc0024e1200, 0x6e9e6a0, 0xc000428, 0x7c59ac8, 0xc002877f80)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/proto/go/provider.pb.go:2194 +0x89
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x7c59ac8, 0xc002877bc0, 0x6e54980, 0xc0024e1200, 0xc0024894c0, 0xc000e574d0, 0x0, 0x0, 0x7b85860, 0xc0003c6c40)
        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x30a
    github.com/pulumi/pulumi/sdk/v3/proto/go._ResourceProvider_Invoke_Handler(0x6fb5c40, 0xc00000ab40, 0x7c59ac8, 0xc002877bc0, 0xc0024e11a0, 0xc001b358a0, 0x7c59ac8, 0xc002877bc0, 0xc0002b4360, 0x102)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v3@v3.0.0/proto/go/provider.pb.go:2196 +0x150
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc00014cc40, 0x7c7d7f8, 0xc001b51500, 0xc002886500, 0xc001c4d620, 0xbfbd400, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1210 +0x52b
    google.golang.org/grpc.(*Server).handleStream(0xc00014cc40, 0x7c7d7f8, 0xc001b51500, 0xc002886500, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:1533 +0xd0c
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc001befd50, 0xc00014cc40, 0x7c7d7f8, 0xc001b51500, 0xc002886500)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:871 +0xab
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.34.0/server.go:869 +0x1f
```

**Comment by stack72** [comment]
Hi @dokuboyejo / @Sodki / @haidaraM 

So, firstly, I am so sorry this has taken so long to get to - this is something we are continually trying to do better with and we will work on this

Secondly, I have tracked down the source of the error here... TL;DR, the use of the apply isn't correct and is what is causing the error. We *should* not be panicing in the application but rather we should be helping point out what the problem is. Given the following example:

```
from pulumi_aws import iam, s3

my_bucket = s3.Bucket(
    "simple-bucket",
    force_destroy=True,
    acl="private",
    versioning=s3.BucketVersioningArgs(enabled=True),
)

policy_document = iam.get_policy_document(
    statements=[
        iam.GetPolicyDocumentStatementArgs(
            sid="",
            effect="Allow",
            actions=["s3:GetObject"],
            resources=[my_bucket.arn.apply(lambda arn: f"{arn}/*")],
        ),
    ]
)
```

we were originally panicing. This is because of the use of the apply inside the get_policy_document. The actual way to use this would be to wrap the Invoke (get_policy_document) with the apply. This program now throws the following error:

```
Diagnostics:
  pulumi:pulumi:Stack (test-py-grpc-dev):
    error: Program failed with an unhandled exception:
    error: Traceback (most recent call last):
      File "/usr/local/bin/pulumi-language-python-exec", line 92, in <module>
        loop.run_until_complete(coro)
      File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
        return future.result()
      File "/Users/stack72/code/test-py-grpc/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 110, in run_in_stack
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/stack72/code/test-py-grpc/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 43, in run_pulumi_func
        func()
      File "/Users/stack72/code/test-py-grpc/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 110, in <lambda>
        await run_pulumi_func(lambda: Stack(func))
      File "/Users/stack72/code/test-py-grpc/venv/lib/python3.8/site-packages/pulumi/runtime/stack.py", line 133, in __init__
        func()
      File "/usr/local/bin/pulumi-language-python-exec", line 91, in <lambda>
        coro = pulumi.runtime.run_in_stack(lambda: runpy.run_path(args.PROGRAM, run_name='__main__'))
      File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py", line 280, in run_path
        return _run_code(code, mod_globals, init_globals,
      File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py", line 86, in _run_code
        exec(code, run_globals)
      File "./__main__.py", line 12, in <module>
        policy_document = iam.get_policy_document(
      File "/Users/stack72/code/test-py-grpc/venv/lib/python3.8/site-packages/pulumi_aws/iam/get_policy_document.py", line 185, in get_policy_document
        __ret__ = pulumi.runtime.invoke('aws:iam/getPolicyDocument:getPolicyDocument', __args__, opts=opts, typ=GetPolicyDocumentResult).value
      File "/Users/stack72/code/test-py-grpc/venv/lib/python3.8/site-packages/pulumi/runtime/invoke.py", line 144, in invoke
        raise invoke_error
    Exception: invoke of aws:iam/getPolicyDocument:getPolicyDocument failed: invocation of aws:iam/getPolicyDocument:getPolicyDocument returned an error:
    Incorrect use of `apply`. Please refer to https://www.pulumi.com/docs/intro/concepts/inputs-outputs/#apply for further details on how to use `apply`
    error: an unhandled error occurred: Program exited with non-zero exit code: 1


```

Notice, no panic and a link to the apply docs - I am going to open a PR there to add a more advanced example like this

The following is the correct way to use this type of Invoke with Apply operation and will do as you expect

```
from pulumi_aws import iam, s3

my_bucket = s3.Bucket(
    "simple-bucket",
    force_destroy=True,
    acl="private",
    versioning=s3.BucketVersioningArgs(enabled=True),
)

policy_document = my_bucket.arn.apply(lambda arn: iam.get_policy_document(
        statements=[
            iam.GetPolicyDocumentStatementArgs(
                sid="",
                effect="Allow",
                actions=["s3:GetObject"],
                resources=[f"{arn}/*"],
            ),
        ]
))

```

Notice the get_policy_document happening inside the apply - this is a valid program and will not error

The PR for the error message is just about to be opened now

Sorry again

Paul

**Comment by t0yv0** [comment]
I am working on a related P1. https://github.com/pulumi/pulumi/issues/7132 at least I am seeing `Unrecognized structpb value` in a different scenario, so adding myself to this, trying to understand it better.

**Comment by stack72** [comment]
This was closed via #7214 and will be available in the next release of Pulumi
